### PR TITLE
add access_points

### PR DESCRIPTION
**Pull request by xlqian**

### CODE REVIEW

File: server/src/response.rs [hunk 1]
```diff
@@ -486,17 +490,98 @@ fn make_stop_point_pt_object(
...
+        .filter_map(|(pathway, access_point_idx)| {
+            make_access_point(model, pathway, access_point_idx)
+        })
+        .collect()
+}
+
 pub fn make_stop_point(
     stop_point_idx: &StopPointIdx,
     model: &ModelRefs,
+    depth: usize,
```
Comment (pbench) [inline server/src/response.rs:584]:
```suggestion
    include_access_points: bool,
```

File: src/models/base_model.rs [hunk 2]
```diff
@@ -77,17 +79,23 @@ pub fn strip_id_prefix<'a>(id: &'a str, prefix: &str) -> &'a str {
...
     loads_data: LoadsData,
     validity_period: (NaiveDate, NaiveDate),
     default_transfer_duration: PositiveDuration,
+    association_stop_point_to_pathways: AssociationStopPoint2PathWays,
 }
 
 pub type BaseVehicleJourneyIdx = Idx<transit_model::objects::VehicleJourney>;
 pub type BaseStopPointIdx = Idx<transit_model::objects::StopPoint>;
 pub type BaseTransferIdx = Idx<transit_model::objects::Transfer>;
+pub type StopLocationIdx = Idx<transit_model::objects::StopLocation>;
```
Comment (pbench) [inline src/models/base_model.rs:98]: This is already defined as AccesPointIdx above. 
We need to decide which one to keep :)

File: src/models/base_model.rs [hunk 3]
```diff
@@ -77,17 +79,23 @@ pub fn strip_id_prefix<'a>(id: &'a str, prefix: &str) -> &'a str {
...
+pub type AccessPointId = Idx<transit_model::objects::StopLocation>;
+pub type AssociationStopPoint2PathWays =
+    HashMap<Idx<transit_model::objects::StopPoint>, HashSet<(PathwayId, AccessPointId)>>;
 
 pub struct BaseModel {
     model: Model,
     loads_data: LoadsData,
     validity_period: (NaiveDate, NaiveDate),
     default_transfer_duration: PositiveDuration,
+    association_stop_point_to_pathways: AssociationStopPoint2PathWays,
```
Comment (pbench) [inline src/models/base_model.rs:92]:
```suggestion
   stop_point_to_pathways: StopPointToPathways,
```

File: src/models/base_model.rs [hunk 4]
```diff
@@ -154,14 +230,20 @@ impl BaseModel {
         let validity_period = model
             .calculate_validity_period()
             .map_err(|_| BadModel::NoDataset)?;
+
         if validity_period.0 > validity_period.1 {
             return Err(BadModel::StartDateAfterEndDate);
         }
+        // Associate stop_points with path way
+        let association_stop_point_to_pathways =
+            Self::associate_stop_points_with_pathway(&model).unwrap();
```
Comment (pbench) [inline src/models/base_model.rs:239]: We don't want to crash the program in the Error case, so we should bubble up the error with something like
```suggestion
            Self::associate_stop_points_with_pathway(&model)
            .map_err(
                |err| BadModel::NewBadModelCase(err)
            )?;
```

File: src/models/base_model.rs [hunk 5]
```diff
@@ -181,6 +263,39 @@ impl BaseModel {
     }
 }
 
+pub struct PathwayByIter<'model> {
+    idx_iter: Option<
+        std::collections::hash_set::Iter<
+            'model,
+            (Idx<Pathway>, Idx<transit_model::objects::StopLocation>),
```
Comment (pbench) [inline src/models/base_model.rs:270]:
```suggestion
            (PathwayIdx, AccessPointIdx),
```

File: src/models/base_model.rs [hunk 6]
```diff
@@ -181,6 +263,39 @@ impl BaseModel {
...
+        let idx_iter = match model.association_stop_point_to_pathways.get(stop_point_idx) {
+            Some(pathways) => Some(pathways.iter()),
+            None => None,
+        };
+        Self { idx_iter, model }
+    }
+}
+
+impl<'model> Iterator for PathwayByIter<'model> {
+    type Item = (&'model Pathway, Idx<transit_model::objects::StopLocation>);
```
Comment (pbench) [inline src/models/base_model.rs:287]:
```suggestion
    type Item = (&'model Pathway, AccessPointIdx);
```

File: src/models/base_model.rs [hunk 7]
```diff
@@ -358,6 +480,33 @@ impl BaseModel {
             None => BTreeSet::new(),
         }
     }
+
+    pub fn get_pathway(&self, stop_point_idx: &BaseStopPointIdx) -> PathwayByIter {
```
Comment (pbench) [inline src/models/base_model.rs:484]:
```suggestion
    pub fn stop_point_pathways(&self, stop_point_idx: &BaseStopPointIdx) -> PathwayByIter {
```

File: src/models/model_refs.rs [hunk 8]
```diff
@@ -415,6 +423,10 @@ impl<'model> ModelRefs<'model> {
             .map(StopPointIdx::Base)
             .collect()
     }
+
+    pub fn get_associated_pathway(&self, stop_point_idx: &BaseStopPointIdx) -> PathwayByIter {
```
Comment (pbench) [inline src/models/model_refs.rs:427]:
```suggestion
    pub fn stop_point_pathways(&self, stop_point_idx: &BaseStopPointIdx) -> PathwayByIter {
```

File: src/models/base_model.rs [hunk 9]
```diff
@@ -358,6 +480,33 @@ impl BaseModel {
...
+    pub fn stop_location_stop_code(&self, idx: StopLocationIdx) -> Option<String> {
+        self.model.stop_locations[idx].code.clone()
+    }
+    pub fn stop_location_coord(&self, idx: StopLocationIdx) -> Coord {
+        let stop_location = &self.model.stop_locations[idx];
+        Coord {
+            lat: stop_location.coord.lat,
+            lon: stop_location.coord.lon,
+        }
+    }
```
Comment (pbench) [inline src/models/base_model.rs:505]: The term access_point is used most of the time to refer to stop_location object, so maybe use access_point here also ?

File: src/models/base_model.rs [hunk 10]
```diff
@@ -77,17 +79,23 @@ pub fn strip_id_prefix<'a>(id: &'a str, prefix: &str) -> &'a str {
 
 pub type Collections = transit_model::model::Collections;
 pub type Model = transit_model::model::Model;
+pub type PathwayId = Idx<transit_model::objects::Pathway>;
+pub type AccessPointId = Idx<transit_model::objects::StopLocation>;
+pub type AssociationStopPoint2PathWays =
+    HashMap<Idx<transit_model::objects::StopPoint>, HashSet<(PathwayId, AccessPointId)>>;
```
[the remark this answers is not in the document]
Comment (xlqian) [reply src/models/base_model.rs:85]: oops, that was a typo, thx

File: src/models/base_model.rs (hunk 2 above)
Comment (xlqian) [reply src/models/base_model.rs:98]: I seem to remember that I did that on purpose. let talk about it later

File: src/models/base_model.rs (hunk 9 above)
Comment (xlqian) [reply src/models/base_model.rs:505]: an `AccessPoint` must be a `StopLocation`, but a `StopLocation` is not necessarily an `AccessPoint`...

File: src/models/base_model.rs [hunk 11]
```diff
@@ -143,9 +150,77 @@ impl BaseModel {
...
+            let (stop_point_id, access_point_id) =
+                if (&pathway.from_stop_type, &pathway.to_stop_type)
+                    == (&StopType::Point, &StopType::StopEntrance)
+                {
+                    (&pathway.from_stop_id, &pathway.to_stop_id)
+                } else if (&pathway.to_stop_type, &pathway.from_stop_type)
+                    == (&StopType::Point, &StopType::StopEntrance)
+                {
+                    (&pathway.to_stop_id, &pathway.from_stop_id)
+                } else {
```
[the remark this answers is not in the document]
Comment (woshilapin) [reply src/models/base_model.rs:189]: As discussed on MP, it's a first approximation that we're OK with. We might come back to that later to support a more general case.

File: server/tests/subtests/places_nearby_test.rs [hunk 12]
```diff
@@ -29,8 +29,52 @@
 
 use loki_server::{navitia_proto, server_config::ServerConfig};
 
+fn check_access_point(
```
[the remark this answers is not in the document]
Comment (xlqian) [reply server/tests/subtests/places_nearby_test.rs:32]: can't compare `&loki_server::navitia_proto::AccessPoint` with `loki_server::navitia_proto::AccessPoint`

File: server/Cargo.toml [hunk 13]
```diff
@@ -40,6 +40,7 @@ serde = "1.0"
 serde_json = "1.0"
 toml = "0.5"
 hostname = "0.3"
+num-traits = "0.2.14"
```
Comment (pbench) [inline server/Cargo.toml:43]: do you use it somewhere ?

File: server/tests/subtests/places_nearby_test.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (woshilapin) [reply server/tests/subtests/places_nearby_test.rs:32]: OK, so the auto-generated protobuf `struct` don't implement it, it's a shame but nothing we can do about it I think.

File: server/src/response.rs [hunk 14]
```diff
@@ -486,17 +489,83 @@ fn make_stop_point_pt_object(
...
+) -> Option<navitia_proto::AccessPoint> {
+    let is_entrance = pathway.is_bidirectional || pathway.from_stop_type == StopType::StopEntrance;
+    let is_exit = pathway.is_bidirectional || pathway.to_stop_type == StopType::StopEntrance;
+
+    let pathway_stop_point_idx = if pathway.from_stop_type == StopType::Point {
+        model.base.stop_point_idx(&pathway.from_stop_id)?
+    } else if pathway.to_stop_type == StopType::Point {
+        model.base.stop_point_idx(&pathway.to_stop_id)?
+    } else {
+        warn!("Something weired occurred when filling pathway: cannot find stop point");
```
Comment (pbench) [inline server/src/response.rs:513]:
```suggestion
        warn!("Unable to convert pathway {} to protobuf : neither from_stop nor to_stop are stop points.", pathway.id);
```

File: server/src/response.rs [hunk 15]
```diff
@@ -486,17 +489,83 @@ fn make_stop_point_pt_object(
...
+        pathway_mode: Some(convert_pathwaymode(&pathway.pathway_mode)),
+        length: pathway.length.map(|x| {
+            x.floor()
+                .to_i32()
+                .or_else(|| {
+                    warn!("cannot convert pathway length to i32");
+                    Some(0)
+                })
+                .unwrap()
+        }),
```
Comment (pbench) [inline server/src/response.rs:539]: Maybe set the length to None instead of Some(0) when the conversion fails

File: server/src/response.rs [hunk 16]
```diff
@@ -486,17 +489,83 @@ fn make_stop_point_pt_object(
...
+        traversal_time: pathway.traversal_time.map(|x| {
+            i32::try_from(x).expect("traversal_time shouldn't be greater than u32::MAX/2")
```
Comment (pbench) [inline server/src/response.rs:541]: Don't panic when the conversion fails ! 
Either set traversal_time to None, or returns from this function with None

File: server/src/response.rs [hunk 17]
```diff
@@ -486,17 +489,83 @@ fn make_stop_point_pt_object(
...
+        }),
+        stair_count: pathway.stair_count.map(|x| x as i32),
+        max_slope: pathway.max_slope.map(|x| x as i32),
+        min_width: pathway.min_width.map(|x| x as i32),
```
Comment (pbench) [inline server/src/response.rs:545]:
```suggestion
        stair_count: pathway.stair_count.map(|x| i32::from(x)),
        max_slope: pathway.max_slope.map(|x| x.floor().to_i32()).flatten(),
        min_width: pathway.min_width.map(|x| x.floor().to_i32()).flatten(),
```
There is a lot of "hidden" things that may happen behind a "as" cast https://doc.rust-lang.org/reference/expressions/operator-expr.html#numeric-cast
So I would rather use more explicit conversion methods

File: server/src/response.rs [hunk 18]
```diff
@@ -524,6 +593,22 @@ pub fn make_stop_point(
     });
 
     let stop_point_name = model.stop_point_name(stop_point_idx);
+
+    let access_points: Vec<navitia_proto::AccessPoint> = if include_access_points {
+        match stop_point_idx {
+            StopPointIdx::Base(base_stop_point_idx) => model
+                .base
+                .stop_point_pathways(base_stop_point_idx)
+                .into_iter()
```
Comment (woshilapin) [inline server/src/response.rs:602]: You can remove that line. `stop_point_pathways()` already return an iterator, so there is no need to convert it to an iterator ([`IntoIterator` is implemented automatically on every `Iterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#impl-IntoIterator-25) but [it basically does nothing](https://doc.rust-lang.org/src/core/iter/traits/collect.rs.html#238-246)).

File: server/Cargo.toml (hunk 13 above)
Comment (pbench) [reply server/Cargo.toml:43]: ok I saw where it is used :+1:

File: src/models/base_model.rs [hunk 19]
```diff
@@ -143,9 +150,77 @@ impl BaseModel {
...
+                    continue;
+                };
+            match Self::insert_pathway_into_association(
+                model,
+                stop_point_id,
+                access_point_id,
+                pathway_idx,
+                &mut association,
+            ) {
+                Some(false) => warn!("Cannot insert pathway to association"),
```
Comment (pbench) [inline src/models/base_model.rs:205]:
```suggestion
                Some(false) => warn!("Inserted twice the pathway {} between stop_point {} and access_point {}",
                    pathway.id,
                    stop_point_id, 
                    access_point_id
                ),
```

File: src/models/base_model.rs [hunk 20]
```diff
@@ -181,6 +261,40 @@ impl BaseModel {
...
+        let (pathway_idx, access_point_idx) = iter.next()?;
+        Some((
+            self.model.model.pathways.index(*pathway_idx),
+            *access_point_idx,
+        ))
+    }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match &self.idx_iter {
+            Some(iter) => iter.size_hint(),
+            None => (0, None),
```
Comment (pbench) [inline src/models/base_model.rs:293]:
```suggestion
            None => (0, 0),
```
You know that the iterator is empty when idx_iter is None

**Review by pbench — APPROVED** [review]

**Pull request merged**